src/Dev/Tasks/FileMigrationHelper.php

```diff
@@ -0,0 +1,351 @@
+        }
+
+        foreach ($this->getLegacyFileQuery() as $file) {
+            // Bypass the accessor and the filename from the column
+            $filename = $file->getField('Filename');
+
+            $success = $this->migrateFile($base, $file, $filename);
+            if ($success) {
+                $ss3Count++;
+            }
+        }
+
+        // Show summary of results
+        if ($ss3Count > 0) {
+            $this->logger->info(sprintf('%d legacy files have been migrated.', $ss3Count));
+        } else {
+            $this->logger->info(sprintf('No SilverStripe 3 files have been migrated.', $ss3Count));
+        }
+
+        if (class_exists(Versioned::class)) {
+            Versioned::set_reading_mode($originalState);
+        }
+
+        return $ss3Count;
+    }
+
+    /**
+     * Migrate a single file
+     *
+     * @param string $base Absolute base path (parent of assets folder)
+     * @param File $file
+     * @param string $legacyFilename
+     * @return bool True if this file is imported successfully
+     */
+    protected function migrateFile($base, File $file, $legacyFilename)
+    {
+        // Make sure this legacy file actually exists
+        $path = $base . '/' . $legacyFilename;
+        if (!file_exists($path)) {
+            return false;
+        }
+
+        // Remove this file if it violates allowed_extensions
+        $allowed = array_filter(File::getAllowedExtensions());
+        $extension = strtolower($file->getExtension());
+        if (!in_array($extension, $allowed)) {
+            if ($this->config()->get('delete_invalid_files')) {
+                $file->delete();
+            }
+            return false;
+        }
+
+        // Fix file classname if it has a classname that's incompatible with it's extention
+        if (!$this->validateFileClassname($file, $extension)) {
+            // We disable validation (if it is enabled) so that we are able to write a corrected
+            // classname, once that is changed we re-enable it for subsequent writes
+            $validationEnabled = DataObject::Config()->get('validation_enabled');
+            if ($validationEnabled) {
+                DataObject::Config()->set('validation_enabled', false);
+            }
+            $destinationClass = $file->get_class_for_file_extension($extension);
+            $file = $file->newClassInstance($destinationClass);
+            $fileID = $file->write();
+            if ($validationEnabled) {
+                DataObject::Config()->set('validation_enabled', true);
+            }
+            $file = File::get_by_id($fileID);
+        }
+
+        // Copy local file into this filesystem
+        $filename = $file->generateFilename();
+        $results = $this->store->normalisePath($filename);
+
+        // Move file if the APL changes filename value
+        $file->File->Filename = $results['Filename'];
+        $file->File->Hash = $results['Hash'];
+
+
+        // Save and publish
+        $file->write();
```

Comment:
Should we wrap this in a try/catch? In my testing, there were files that didn't meet the allowed_extension requirements that had gotten into the system one way or another. Recommend we capture the validation exception and log it out.